### PR TITLE
Pdga 388  implement IDatabaseModel interface

### PR DESCRIPTION
**Pull request by drew-acheson**
This will be followed up by a branch in php-api.

See [https://pdga.atlassian.net/browse/PDGA-388](https://pdga.atlassian.net/browse/PDGA-388)

### CODE REVIEW

**Review by pcrist — COMMENTED** [review]
This makes sense to me; should the removed tests be updated to use `IDatabaseModel` instead, or are we already testing elsewhere?

**Comment by drew-acheson** [comment]
> This makes sense to me; should the removed tests be updated to use `IDatabaseModel` instead, or are we already testing elsewhere?

As far as I can tell those tests were testing the array centric branch of the code which isn't used now that the new interface has been created.

**Review by pcrist — APPROVED** [review]

**Comment by drew-acheson** [comment]
@benbotto When testing this locally with php-api I got errors from the tests. Apparently Laravel's Models don't explicitly return arrays which keeps them from being compatible with the interface the way it's defined.  Removing the `:array` from the function definition fixes the issue. Does this seem counter to the explicit reason for this task?

File: src/Models/ModelInstantiator.php [hunk 1]
```diff
@@ -148,15 +149,14 @@ public function dataObjectToDatabaseModel(
...
      *
      * @throws ReflectionException
      * @return object
      */
     public function databaseModelToDataObject(
-        $db_model,
+        IDatabaseModel $db_model,
         string $class
     ): object
     {
```
[the remark this answers is not in the document]
Comment (drew-acheson) [reply src/Models/ModelInstantiator.php:162]: Can do.  Did you see the comment above about removing the return types from the inteface?

**Comment by bbotto-pdga** [comment]
> Laravel's Models don't explicitly return arrays which keeps them from being compatible with the interface the way it's defined.

That's a shame.  For reference, here's the relevant Eloquent code:

https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Eloquent/Concerns/HasRelationships.php#L816
https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L1791

> Removing the `:array` from the function definition fixes the issue. Does this seem counter to the explicit reason for this task?

No, I don't think it goes against the reasons for this change.  One reason is optimization.  It's obviously faster to use models directly than to convert the model to an array, but checking for the `getAttributes`/`getRelations` methods was a 2.x backwards-compatible change that I hacked in.  With the interface, we can expect those methods to be there.

The other reason is that we want to accept Eloquent models, but we don't want `php-data-objects` to have a Composer dependency on Eloquent because of the upkeep that would be required.  The interface allows us to be explicit about which Model methods are used without the dependency.

Without the type hints on the base class, we can't _technically_ be guaranteed that the methods return arrays, but we _practically_ can, if that makes sense.

Any input, @timallen527?

**Comment by timallen527** [comment]
@bbotto-pdga I agree that it does not go against the original reason for the change. It is unfortunate that the return types aren't explicit on the Laravel side. That handcuffs us a bit as far as what we can do in terms of type safety. 

The only idea to work around this that comes to mind initially is to create some type of pass-through abstraction (facade pattern or something similar?) that sits on top of `\Illuminate\Database\Eloquent\Model` that would give us the ability to be explicit in the typing. The idea being that this use this abstraction in `php-data-objects` and it feels and acts just like the original Model because it just passes any calls through to the original. But, we get control over the method signatures. 

I think this approach would work, but I don't think we should go that way for a couple reasons. The implementations would have to be scattered across all projects that want to use `php-data-objects` to avoid the Eloquent dependency problem. This would also require a bit of maintained code that would create barriers to doing things like updating Eloquent.

File: src/Interfaces/IDatabaseModel.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+     * @return array The attributes of the database model.
+     */
+    public function getAttributes(): array;
+
+    /**
+     * Get the relations of the database model.
+     *
+     * @return array The relations of the database model.
+     */
+    public function getRelations(): array;
```
Comment (bbotto-pdga) [inline src/Interfaces/IDatabaseModel.php:18]: Don't the return types needed to be removed from these two methods?

Comment (drew-acheson) [reply src/Interfaces/IDatabaseModel.php:18]: Yes, I was just waiting for a consensus and then I was working on something else.  I'll do that and close this today and open the php-api branch that uses this.

**Review by bbotto-pdga — APPROVED** [review]

**Pull request merged**